pkg/controller/helper/helper_test.go

```diff
@@ -0,0 +1,181 @@
+package helper
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"runtime"
+	"testing"
+
+	"github.com/3scale/3scale-operator/pkg/helper"
+
+	threescaleapi "github.com/3scale/3scale-porta-go-client/client"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// assert fails the test if the condition is false.
+func assert(tb testing.TB, condition bool, msg string, v ...interface{}) {
+	if !condition {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d: "+msg+"\033[39m\n\n", append([]interface{}{filepath.Base(file), line}, v...)...)
+		tb.FailNow()
+	}
+}
+
+// ok fails the test if an err is not nil.
+func ok(tb testing.TB, err error) {
+	if err != nil {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d: unexpected error: %s\033[39m\n\n", filepath.Base(file), line, err.Error())
+		tb.FailNow()
+	}
+}
+
+// equals fails the test if exp is not equal to act.
+func equals(tb testing.TB, exp, act interface{}) {
+	if !reflect.DeepEqual(exp, act) {
+		_, file, line, _ := runtime.Caller(1)
+		fmt.Printf("\033[31m%s:%d:\n\n\texp: %#v\n\n\tgot: %#v\033[39m\n\n", filepath.Base(file), line, exp, act)
+		tb.FailNow()
+	}
+}
+
+func NewTestAdminPortal(t *testing.T) *threescaleapi.AdminPortal {
+	t.Helper()
+	ap, err := threescaleapi.NewAdminPortalFromStr("https://www.test.com:443")
+	ok(t, err)
+	return ap
+}
+
+func GetTestSecret(namespace, secretName string, data map[string]string) *v1.Secret {
+	secret := &v1.Secret{
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Secret"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: namespace,
+		},
+		StringData: data,
+		Type:       v1.SecretTypeOpaque,
+	}
+	secret.Data = helper.GetSecretDataFromStringData(secret.StringData)
+	return secret
+}
+
+// RoundTripFunc .
+type RoundTripFunc func(req *http.Request) *http.Response
```

Comment:
Go includes a RountTripper interface which I think is intended to do the same: https://golang.org/pkg/net/http/#RoundTripper

Would it be a better idea to reuse that?

Reply:
That interface is what `RoundTripFunc` type implements. It needs to be like that to be assigned in the `Transport` field of the http client. 

This `RoundTripFunc` makes the interface to the test users a little bit easier to use. Instead of
```
RoundTrip(*Request) (*Response, error)
```
the test user must implement
```
func(req *http.Request) *http.Response
```
and if there is an error, the test should raise the error. For instance:
```
httpClient := NewTestClient(func(req *http.Request) *http.Response {              
    respObject := struct {                                                        
        Errors map[string][]string `json:"errors"`                                
    }{                                                                            
        map[string][]string{"some_attr": []string{"not valid"}},                  
    }                                                                             
                                                                                  
    responseBodyBytes, err := json.Marshal(respObject)                            
    ok(t, err)             <- raise test failed if there is error                                                        
                                                                                  
    return &http.Response{                                                        
        StatusCode: http.StatusUnprocessableEntity,                               
        Body:       ioutil.NopCloser(bytes.NewBuffer(responseBodyBytes)),         
        Header:     make(http.Header),                                            
    }                                                                             
})                                                                                
```
What do you have in mind?

Reply:
I see.

Thanks for the explanation. This approach looks good to me. It allows the user not having to create its own type and provide a method instead.